File: MATLAB_models/Subsurface/WorkingScript.m
                % ============================ %
                % GeoTES Subsurface Cost Model %
                % ============================ %

% Working script for GeoTES subsurface costs. The includes capital costs for
% exploration, permitting, drilling & completionss, pumping, surface flow
% lines. It also calculates the subsurface O&M cost, excluding labor cost.

% The input arguments in the defined classes are placeholders. Users should
% define their own arguments based on the project needs.


% =========================
% Capital Cost Calculations
% =========================

% Exploration Cost
Exploration_Cost = ExplorationCost("Greenfield",3,20000,1.2);
Total_Exploration_Cost = Exploration_Cost.ExplorationCostCalc();
disp("Total Exploration Cost = $" + Total_Exploration_Cost);

% Development and Plant Permitting
PPImultiplier = 1.175; % 2012 dollar year (for legal services)
FieldDevPlantPermitting = 1000000 * PPImultiplier; % from GETEM 2012 cost

% Drilling cost and completions
Drilling_Cost = DrillingCost(4,5,2500,500,"Large",1,2.2,"Vertical","Openhole");
Well_Cost = Drilling_Cost.WellCost();
Total_Drilling_Cost = Drilling_Cost.TotalDC(Well_Cost);
disp("Total Drilling Cost = $" + Total_Drilling_Cost);

% Production pumping duty and cost
ProdPumping_Estimation = ProductionPumpingCost(50,110,4,2500,2500,500,"Large","Lineshaft","Openhole",1.553);
Step1 = ProdPumping_Estimation.HeadProdTop();
Step2 = ProdPumping_Estimation.HeadSunction(Step1.Prodtop);
Step3 = ProdPumping_Estimation.Sunctiondepth(Step2);
Step4 = ProdPumping_Estimation.CasingFriction(Step3,Step1.Prodtop);
Step5 = ProdPumping_Estimation.Pumppower(Step3,Step4);
Step6 = ProdPumping_Estimation.Pumpcost(Step5,Step3);
ProdPumping_Cost_Duty = ProdPumping_Estimation.TotalPumpDutyCost(Step5,Step6);

% Injection pumping cost
InjPumping_Estimation = InjectionPumpingCost(50,160,146.67,110,187.25,2.5,3000,2500,500,"Large","Openhole",1.533,"Discharge");
Step7 = InjPumping_Estimation.HeadSunction();
Step8 = InjPumping_Estimation.HeadInjection(Step7);
Step9 = InjPumping_Estimation.Pumppower(Step8);
Step10 = InjPumping_Estimation.Pumpcost(Step9);
InjPumping_Cost_Duty = InjPumping_Estimation.TotalPumpDutyCost(Step9,Step10);

% Flow line cost. This assumes one pad per well. Will include estimation
% for multi-well pads after discussion with PRM and Earthbridge.
LengthofFlowline = 750; % in meters per well (GETEM assumption)
PipingUnitCost = 256.98; % $ per ft (in 2022 dollars)
TotalNumberofWells = 11; % User-defined
FlowLineCost = LengthofFlowline * 3.28084 * PipingUnitCost * TotalNumberofWells;

% Add fluid Processing bulk cost for oilfield from PRM


% Total Subsurface Capital Cost
% =============================
SubsurfaceCapitalCost = Total_Exploration_Cost + FieldDevPlantPermitting + ...
    Total_Drilling_Cost + Step6 + Step10 + FlowLineCost;
disp("Subsurface Capital Cost = $" + SubsurfaceCapitalCost);



% =====================
% O&M Cost Calculations
% =====================

% This excludes labor cost which will be determined from the plant gross
% output

% Wellfield Maintenance
WellFieldMaintenance = 0.015 * (Total_Drilling_Cost + FlowLineCost); % 1.5% of Wellfield cost 

% Pump Maintenance
PumpMaintenance = ProdPumping_Estimation.PumpMaintenanceCost(Step3,Step6);

% Makeup Water Cost
OilSaturation = 0; % This defaults to an aquifer. For an oil reservoir, the user should input the avg. oil saturation
NumberProductionWells = 4; % The user should change this based on number of production wells
ProductionRateperWell = 110; % in kg/s. The user should change this based plant design
MakeupWaterUnitCost = 0.65; % $ per bbl or 5818.50 per arce-ft.
% (https://waterstandard.com/the-role-of-produced-water-treatment-in-shale-plays/).
% $300 per acre-ft in GETEM
SubsurfaceWaterLoss = OilSaturation + 0.05;
MakeupWaterSubsurface = MakeupWaterUnitCost * NumberProductionWells * ...
    (ProductionRateperWell * 18.0917 * 34.2857 * 365) * SubsurfaceWaterLoss; % in $ per year

% Property Tax and Insurance
PropertyTaxRate = 0.0075; % 0.75% rate from GETEM
AnnualTaxandInsurance = PropertyTaxRate * SubsurfaceCapitalCost;


% Total Subsurface O&M Cost
% =========================
SubsurfaceOandMCost = WellFieldMaintenance + PumpMaintenance + MakeupWaterSubsurface + AnnualTaxandInsurance;
disp("Subsurface O&M Cost = $" + SubsurfaceOandMCost);

File: Python_TEA_model/Subsurface TEA Model/WorkingScript.py
from ExplorationCost import ExplorationCost
from DrillingCost import DrillingCost
from ProductionPumpingCost import ProductionPumpingCost
from InjectionPumpingCost import InjectionPumpingCost


def main():
    exploration_cost = ExplorationCost("Greenfield", 3, 20000, 1.2)
    total_exploration_cost = exploration_cost.exploration_cost_calc()
    print(f"Total Exploration Cost = ${total_exploration_cost}")

    ppi_multiplier = 1.175
    field_dev_plant_permitting = 1_000_000.0 * ppi_multiplier

    drilling_cost = DrillingCost(
        4,
        5,
        2500,
        500,
        "Large",
        1.0,
        2.2,
        "Vertical",
        "Openhole",
    )
    well_cost = drilling_cost.well_cost()
    total_drilling_cost = drilling_cost.total_dc(well_cost)
    print(f"Total Drilling Cost = ${total_drilling_cost}")

    prod_pumping = ProductionPumpingCost(
        190,
        110,
        4,
        2500,
        2500,
        500,
        "Large",
        "Lineshaft",
        "Openhole",
        1.553,
    )
    step1 = prod_pumping.head_prod_top()
    step2 = prod_pumping.head_suction(step1)
    step3 = prod_pumping.suction_depth(step2)
    step4 = prod_pumping.casing_friction(step3, step1)
    step5 = prod_pumping.pump_power(step3, step4)
    step6 = prod_pumping.pump_cost(step5, step3)
    prod_pumping_cost_duty = prod_pumping.total_pump_duty_cost(step5, step6)

    inj_pumping = InjectionPumpingCost(
        190,
        70,
        146.67,
        110,
        187.25,
        2.5,
        3000,
        2500,
        500,
        "Large",
        "Openhole",
        1.533,
        "Discharge",
    )
    step7 = inj_pumping.head_suction()
    step8 = inj_pumping.head_injection(step7)
    step9 = inj_pumping.pump_power(step8)
    step10 = inj_pumping.pump_cost(step9)
    inj_pumping_cost_duty = inj_pumping.total_pump_duty_cost(step9, step10)

    length_of_flowline = 750.0
    piping_unit_cost = 256.98
    total_number_of_wells = 11
    flow_line_cost = length_of_flowline * 3.28084 * piping_unit_cost * total_number_of_wells

    subsurface_capital_cost = (
        total_exploration_cost
        + field_dev_plant_permitting
        + total_drilling_cost
        + step6
        + step10
        + flow_line_cost
    )
    print(f"Subsurface Capital Cost = ${subsurface_capital_cost}")

    wellfield_maintenance = 0.015 * (total_drilling_cost + flow_line_cost)
    pump_maintenance = prod_pumping.pump_maintenance_cost(step3, step6)

    oil_saturation = 0.0
    number_production_wells = 4
    production_rate_per_well = 110.0
    makeup_water_unit_cost = 0.65
    subsurface_water_loss = oil_saturation + 0.05
    makeup_water_subsurface = (
        makeup_water_unit_cost
        * number_production_wells
        * (production_rate_per_well * 18.0917 * 34.2857 * 365.0)
        * subsurface_water_loss
    )

    property_tax_rate = 0.0075
    annual_tax_and_insurance = property_tax_rate * subsurface_capital_cost

    subsurface_om_cost = (
        wellfield_maintenance
        + pump_maintenance
        + makeup_water_subsurface
        + annual_tax_and_insurance
    )
    print(f"Subsurface O&M Cost = ${subsurface_om_cost}")


if __name__ == "__main__":
    main()
